marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/ConstraintsSpec.hs

```diff
@@ -0,0 +1,734 @@
+                      ("Expected to consume UTxO " <> show txOutRef)
+                  , check
+                      (any ((== transactionOutput) . fromCardanoTxOut BabbageEra) txOuts)
+                      ("Matching output not found for input " <> show transactionOutput)
+                  ]
+      _ -> []
+
+mustPayToAddressViolations
+  :: MarloweVersion v -> TxConstraints v -> TxBodyContent BuildTx BabbageEra -> [String]
+mustPayToAddressViolations MarloweV1 TxConstraints{..} TxBodyContent{..} = do
+  (address, assets) <- Map.toList payToAddresses
+  (("address: " <> show address <> ": ") <>) <$> do
+    let
+      totalToAddress = foldMap extractValue
+        $ filter ((== address) . extractAddress) txOuts
+    check
+      (totalToAddress == assets)
+      ("Address paid the wrong amount. Expected " <> show assets <> " got " <> show totalToAddress)
+
+mustSendMarloweOutputViolations
+  :: MarloweVersion v
+  -> MarloweContext v
+  -> TxConstraints v
+  -> TxBodyContent BuildTx BabbageEra
+  -> [String]
+mustSendMarloweOutputViolations MarloweV1 MarloweContext{..} TxConstraints{..} TxBodyContent{..} =
+  case marloweOutputConstraints of
+    MarloweOutputConstraintsNone -> []
+    MarloweOutput assets datum ->
+      case find ((== marloweAddress) . extractAddress) txOuts of
+        Nothing -> ["No output found to Marlowe address"]
+        Just txOut -> fold
+          [ check (extractValue txOut == assets) "Wrong assets sent to Marlowe Address."
+          , check (extractDatum txOut == Just (toChainDatum MarloweV1 datum)) "Wrong assets sent to Marlowe Address."
+          ]
+
+mustSendMerkleizedContinuationOutputViolations
+  :: MarloweVersion v -> TxConstraints v -> TxBodyContent BuildTx BabbageEra -> [String]
+mustSendMerkleizedContinuationOutputViolations MarloweV1 TxConstraints{..} TxBodyContent{..} = do
+  continuationContract <- Set.toList merkleizedContinuationsConstraints
+  let isMatch = (== Just (Chain.toDatum continuationContract))
+  let matchingOutput = find (isMatch . extractDatum) txOuts
+  case matchingOutput of
+    Nothing -> ["No matching output found"]
+    _ -> []
+
+mustPayToRoleViolations
+  :: MarloweVersion v
+  -> MarloweContext v
+  -> TxConstraints v
+  -> TxBodyContent BuildTx BabbageEra
+  -> [String]
+mustPayToRoleViolations MarloweV1 MarloweContext{..} TxConstraints{..} TxBodyContent{..} = do
+  (roleToken, assets) <- Map.toList payToRoles
+  (("roleToken" <> show roleToken <> ": ") <>) <$> do
+    let
+      isMatch txOut = extractAddress txOut == payoutAddress
+                   && extractDatum txOut == Just (toChainPayoutDatum MarloweV1 roleToken)
+      matchingOutput = find isMatch txOuts
+    case matchingOutput of
+      Nothing -> ["No matching output found to payout validator"]
+      Just txOut ->
+        let
+          totalToRole = extractValue txOut
+        in
+          check
+            (totalToRole == assets)
```

Comment:
Should this be an inequality? What about the case when the lovelace in the output is rounded up to satisfy min-UTxO?

Reply:
`solveInitialTxBody` is run pre-minUTxO adjustment (That's why we wanted to write these tests for that function rather than `solveConstraints`, so we would have easier invariants to check).